### ISSUE TITLE
Support both Attribute.name and AttributeOverride.rename 

### ISSUE THREAD

**Issue by getglad** [python-attrs/cattrs#347]
### TLDR 

I'd like to have a `rename` ergonomic such as `override(rename="<FieldName>", accept_both_names=True)` that supports both the actual attribute as well as the `rename` mapping. I have a working implementation that I could try to convert into a PR if this strategy fits.

### Description

1. I have a scenario where I will have a set of input key names that will be changed, and for some period of time I expect to have have both formats in my system. I will still need to be able to `structure` both of them into an `attr`.
2. I would like to be able to convert the `attr` into the new format and make all of my internal changes without doing much `post_init` work in the `attr` (ie, populating an old field into a new field if old field was used).

In a simple example, my format will convert from `{'OldField':  'value1'}` to `{'NewField': 'value2'}`.

And my `attr` would change from

```
@define
class MyInternalAttr:
    OldField: str
```
to 

```
@define
class MyInternalAttr:
    NewField: str
```

If I only needed to cast `OldField` -> `NewField`, I know I could use `make_dict_structure_fn` with  `NewField=override(rename="OldField")`, but at present I could not structure both `OldField` and `NewField`, at least not on the same converter.

My goal is along the lines of what [rename](https://catt.rs/en/stable/customizing.html#rename) can already support, except that rename operates as a full override - I can have the old field or the new field but not both.

### Suggested Implementation

What I'd like to have is an ergonomic along the lines of `NewField=override(rename="OldField", accept_both_names=True)`

I have some stubbed out changes that seem to work, but wanted to check in before I worked towards a PR. The below is a summary of changes I had to make in `cattrs.gen`.

#### Update `AttributeOverride` and the `override` method

```
@frozen
class AttributeOverride:
    omit_if_default: Optional[bool] = None
    rename: Optional[str] = None
    omit: bool = False  # Omit the field completely.
    accept_both_names: bool = False


def override(
    omit_if_default: Optional[bool] = None,
    rename: Optional[str] = None,
    omit: bool = False,
    accept_both: bool = False,
):
    return AttributeOverride(
        omit_if_default=omit_if_default,
        rename=rename,
        omit=omit,
        accept_both_names=accept_both,
    )
```

##### Example in Current Release

https://github.com/python-attrs/cattrs/blob/405f0291b958ae9eb45ee38febeb91fb65dd644f/src/cattrs/gen.py#L336-L350

##### Code Change

I recognize this would need to be added in a few locations, but I think the change ends up being similar everywhere.

```
if a.default is not NOTHING:
  check = f"{i}if '{kn}' in o"
  if override.accept_both_names:
    check += f" or '{an}' in o"
  lines.append(f"{check}:")
  i = f"{i}  "
  lines.append(f"{i}try:")
  i = f"{i}  "
  if handler:
    if handler == converter._structure_call:
      internal_arg_parts[struct_handler_name] = t
      lines.append(f"{i}res['{ian}'] = {struct_handler_name}(o['{kn}'])")
    else:
      type_name = f"__c_type_{an}"
      internal_arg_parts[type_name] = t
      lines.append(f"{i} if '{kn}' in o:")
      lines.append(
        f"{i}{i}res['{ian}'] = {struct_handler_name}(o['{kn}'], {type_name})"
      )
      if override.accept_both_names:
        lines.append(f"{i} elif '{an}' in o:")
        lines.append(
          f"{i}{i}res['{ian}'] = {struct_handler_name}(o['{an}'], {type_name})"
        )
```

The above enables me to do the below

```
converter = Converter()

def support_both_formats(cls):
    return custom_make_dict_structure_fn(
        cls,
        converter,
        NewField=custom_override(rename="OldField", accept_both_names=True),
    )


converter.register_structure_hook_factory(has, support_both_formats)

converter.structure({"NewField": "foo"}, MyInternalAttr)
> MyInternalAttr(NewField='foo'])
converter.structure({"OldField": "foo"}, MyInternalAttr)
> MyInternalAttr(NewField='foo'])
```

**Comment by Tinche** [comment]
Howdy!

I think this is a cool idea but I think this can be done differently, without changes to `gen` and with a little more composability.

How bout something like the following:

```python
from attrs import define

from cattrs import Converter
from cattrs.gen import make_dict_structure_fn

converter = Converter()


def fallback_field(
    converter: Converter, cls: type, old_to_new_field: dict[str, str]
) -> None:
    stuct = make_dict_structure_fn(cls, converter)

    def structure(d: dict, cl):
        for k, v in old_to_new_field.items():
            if k in d:
                d[v] = d[k]

        return stuct(d, cl)

    converter.register_structure_hook(cls, structure)


@define
class MyInternalAttr:
    NewField: str


fallback_field(converter, MyInternalAttr, {"OldField": "NewField"})

print(converter.structure({"NewField": "foo"}, MyInternalAttr))
print(converter.structure({"OldField": "foo"}, MyInternalAttr))
```

The main idea here is you can wrap the generated handling in a little bit of preprocessing.

If this is good for you and you still want to contribute, I've love to get a section for this in the Usages part of the docs.

**Comment by getglad** [comment]
Oh that's an interesting approach - let me play with it and I'll report back. 

Are you thinking an update to the top level usage page (https://catt.rs/en/stable/usage.html) or elsewhere?

**Comment by Tinche** [comment]
Yeah that's the one. I'd like for that to grow into a collection of useful recipes for the community.

**Comment by getglad** [comment]
@Tinche - I imagine you will get the GH notification, but to close the loop - opened https://github.com/python-attrs/cattrs/pull/351 - we can move the conversation there if you like

**Comment by getglad** [comment]
Closing as #351 was merged
